src/main/java/io/pinecone/PineconeDataPlaneClient.java

```diff
@@ -0,0 +1,222 @@
+        QueryRequest.Builder queryRequest = QueryRequest.newBuilder();
+
+        if (id != null && !id.isEmpty() && vectors != null && !vectors.isEmpty()) {
+            throw new PineconeValidationException("Invalid Query Request. Both Vector ID and vectors cannot be present.");
+        }
+
+        if (id != null && !id.isEmpty()) {
+            queryRequest.setId(id);
+        }
+
+        if (namespace != null && !namespace.isEmpty()) {
+            queryRequest.setNamespace(namespace);
+        }
+
+        queryRequest.setTopK(topK)
+                .setIncludeValues(includeValues)
+                .setIncludeMetadata(includeMetadata);
+
+        if (filter != null) {
+            queryRequest.setFilter(filter);
+        }
+
+        if (vectors != null && !vectors.isEmpty()) {
+            queryRequest.addAllVector(vectors);
+        }
+
+        if (sparseIndices != null && sparseValues != null) {
+            if (sparseIndices.size() != sparseValues.size()) {
+                throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both sparse indices and values are of the same length.");
+            }
+
+            queryRequest.setSparseVector(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        return blockingStub.query(queryRequest.build());
+    }
+
+    public FetchResponse fetch(List<String> ids,
+                               String namespace) {
+        FetchRequest.Builder fetchRequest = FetchRequest.newBuilder();
+
+        if (ids == null || ids.isEmpty()) {
+            throw new PineconeValidationException("Invalid Fetch Request. Vector IDs must be present");
+        }
+
+        fetchRequest.addAllIds(ids);
+
+        if (namespace != null && !namespace.isEmpty()) {
+            fetchRequest.setNamespace(namespace);
+        }
+
+        return blockingStub.fetch(fetchRequest.build());
+    }
+
+    public UpdateResponse update(String id,
+                                 List<Float> values,
+                                 Struct metadata,
+                                 String namespace,
+                                 List<Long> sparseIndices,
+                                 List<Float> sparseValues) {
+        UpdateRequest.Builder updateRequest = UpdateRequest.newBuilder();
+
+        if (id == null) {
+            throw new PineconeValidationException("Invalid Update Request. Vector ID must be present");
+        }
+        updateRequest.setId(id);
+
+        if (values != null) {
+            updateRequest.addAllValues(values);
+        }
+
+        if (metadata != null) {
+            updateRequest.setSetMetadata(metadata);
+        }
+
+        if (namespace != null && !namespace.isEmpty()) {
+            updateRequest.setNamespace(namespace);
+        }
+
+        if (sparseIndices != null && sparseValues != null) {
```

Comment:
Similar to query, we may also need to worry about the error cases where only one of `sparseIndices` or `sparseValues` is set. This looks like if `sparseIndices == null` you would skip this check, then hit a NPE when trying to do the int/long conversion.

Reply:
Currently, I'm skipping them so I'll add a validation that if sparseIndices is not null, then the sparseValues should be not null and of the same length

Reply:
Updated the logic and added a test. Will check for if either one is not null, then throw an error. If both are null, continue. Lastly, continue with both not null for checking same size() and then setting it up.